### PR TITLE
feat: load policy categories only when needed

### PR DESCRIPTION
**Pull request by jaydamani**
### Details

policy categories were not loaded on app start due to which the 'categorise it' button did not show policies that the user did not visit after login
### Fixed Issues

$ https://github.com/Expensify/App/issues/47854
PROPOSAL: https://github.com/Expensify/App/issues/47854#issuecomment-2305583095

### Tests

1. Go to staging.new.expensify.com
2. Create a few workspaces
3. Go to self DM
4. Send a message
5. Log out and log it
6. Do not open any workspace chat and go to self DM
7. Click + > Track expense > Manual
8. Submit a track manual expense
9. Click Categorize it
10. Search for other existing workspace chats that do not show up in the list
11. Note that the other existing workspace chats do not return in the search result
12. Open all the workspace chats
13. Repeat Step 9 and 10
14. Note that workspace chats are only searchable after the workspace chats are visited

- [x] Verify that no errors appear in the JS console

### QA Steps

1. Go to staging.new.expensify.com
2. Create a few workspaces
3. Go to self DM
4. Send a message
5. Log out and log it
6. Do not open any workspace chat and go to self DM
7. Click + > Track expense > Manual
8. Submit a track manual expense
9. Click Categorize it
10. Search for other existing workspace chats that do not show up in the list
11. Note that the other existing workspace chats do not return in the search result
12. Open all the workspace chats
13. Repeat Step 9 and 10
14. Note that workspace chats are only searchable after the workspace chats are visited

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/user-attachments/assets/83d95e75-68aa-40c4-8c22-a9eb43bf78cf

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/user-attachments/assets/a856eaa2-1924-48ee-bab7-22aac2463004

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/user-attachments/assets/26631f63-ca73-4eae-86b0-c634d5cf6075

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/user-attachments/assets/498dcbf2-fe8e-4178-b1e6-ea7a6f53cdb2

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/user-attachments/assets/97f46fd9-1f55-4aa4-8b88-652251e6547c

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/user-attachments/assets/68f72d5d-045f-4d24-a9a5-3b8d5b715308

</details>

### CODE REVIEW

**Comment by jaydamani** [comment]
For offline mode after the user selects a workspace, since no categories are loaded for workspace it will show like below instead. This seems incorrect as the workspace does have categories but they are not loaded because user is offline. Should we just show a message that says you need to be online to view this or do we have a common component for this?

image

**Comment by jaydamani** [comment]
There might be an oversight. There is already an API [call](https://github.com/Expensify/App/blob/965bcf9c5a6e90e530d137fdde600ebd9124b7d5/src/pages/iou/request/step/IOURequestStepCategory.tsx#L111) which loads the categories when user selects the policy so, the new endpoint may not be needed.

**Comment by jaydamani** [comment]
@dukenv0307 As part of the [solution](https://github.com/Expensify/App/issues/47854#issuecomment-2305583095) for the orignal issue, we created a new endpoint getPolicyCategories from backend. This PR updates frontend to use the new endpoint for the "categorise it" flow to load all categories after the user selects a workspace but while making the changes from review I noticed that "categorise it" flow already calls another endpoint from [here](https://github.com/Expensify/App/blob/965bcf9c5a6e90e530d137fdde600ebd9124b7d5/src%2Fpages%2Fiou%2Frequest%2Fstep%2FIOURequestStepCategory.tsx#L111) which loads required data related to workspace including the categories for that workspace. So we may not need new endpoint to load categories from the CategoryPicker.
cc: @thienlnam

**Comment by thienlnam** [comment]
The new endpoint is the same as openDraftWorkspaceRequest, it's an alias so the request is 1:1:1 for this different use-case. 

Also, I agree we will need to update this offline pattern since you cannot categorize offline so probably the OfflineIndicator

**Comment by jaydamani** [comment]
@dukenv0307 This is now ready for review.
Here is preview for when user is offline:
image
image

**Comment by jaydamani** [comment]
@dukenv0307 I also agree that looks more appropriate. should we keep it consistent and block the whole flow so show full page offline view when user clicks on "categorize it" or only show it when user selects a workspace and we do not have data for that workspace.
Current offline flow:
- User clicks on categorise it
- User selects a workspace for the expense
- If categories are already loaded, we show the stale data
- If categories are not loaded we show an empty state asking user to create categories like [this](https://github.com/Expensify/App/pull/49756#issuecomment-2375460507) but with offline indicator at bottom

**Comment by trjExpensify** [comment]
That makes sense to me!

**Comment by dubielzyk-expensify** [comment]
Yep that looks good to me

File: src/pages/iou/request/step/StepScreenWrapper.tsx [hunk 1]
```diff
@@ -25,6 +25,9 @@ type StepScreenWrapperProps = {
     /** Whether or not to display not found page */
     shouldShowNotFoundPage?: boolean;
 
+    /** Whether to show offline indicator */
+    shouldShowOfflineIndicator?: boolean;
```
[the remark this answers is not in the document]
Comment (jaydamani) [reply src/pages/iou/request/step/StepScreenWrapper.tsx:29]: StepScreenWrapper renders ScreenWrapper which shows OfflineInidicator at bottom for small screen but our page would already show that user is offline so showing an indicator at bottom would feel weird.
https://github.com/Expensify/App/blob/9627f5f8da6542b42a7f2eb9125dc82eea083740/src/components/ScreenWrapper.tsx#L123
https://github.com/Expensify/App/blob/9627f5f8da6542b42a7f2eb9125dc82eea083740/src/components/ScreenWrapper.tsx#L288

Basically, to avoid this:
image
 
Another approach I tried was to add below code to IOURequestStepCategory but then the header will not render because it comes from StepScreenWrapper so the current approach seemed better
```js
if (policyCategories === undefined && isOffline) {
  // a children is mandatory for FullPageOfflineBlockingView
  return <FullPageOfflineBlockingView>{null}</FullPageOfflineBlockingView>
}
```

**Comment by jaydamani** [comment]
> @jaydamani Can you take a look at my concern above?

I was away yesterday, will work on it and push today.

**Comment by jaydamani** [comment]
@dukenv0307 you can review it

**Review by dukenv0307 — APPROVED** [review]

**Review by thienlnam — APPROVED** [review]
